docs/source/daml/intro/9_Functional101.rst

```diff
@@ -0,0 +1,492 @@
+  :language: daml
+  :start-after: -- BOOL_TO_INT_BEGIN
+  :end-before: -- BOOL_TO_INT_END
+
+If you write this function in the IDE, you'll get a warning from the linter:
+
+.. code-block::none
+
+    Suggestion: Use if
+    Found:
+    case b of
+        True -> 1
+        False -> 0
+    Perhaps:
+    if b then 1 else 0
+
+The linter knows the equivalence and suggests a better implementation:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- BOOL_TO_INT2_BEGIN
+  :end-before: -- BOOL_TO_INT2_END
+
+In short: ``if..else`` statements are equivalent to a ``case`` statement, but are easier to read.
+
+Control Flow as Expressions
+~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+``case`` statements and ``if..else`` really are control flow in the sense that they short circuit:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- TEXT_TO_BOOL_BEGIN
+  :end-before: -- TEXT_TO_BOOL_END
+
+This function behaves as you expect. The error only gets evaluated if an invalid text is passed in.
+
+This is different to functions, where all arguments are evaluated immediately:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- IFELSE_BEGIN
+  :end-before: -- IFELSE_END
+
+In the above, ``boom`` is an error.
+
+But while being proper control flow, ``case`` and ``if..else`` statements are also expressions in the sense that they result in a value when evaluated. You can actually see that in the function definitions above. Since each of the functions is defined just as a ``case`` or ``if`` statement, the value of the evaluated function is just the value of the ``case``/``if`` statement. Things that have a value have a type. The ``if..else`` expression in ``boolToInt2`` has type ``Int`` as that's what the function returns, the ``case`` expression in ``doError`` has type ``Bool``. To be able to give such expressions an unambiguous type, each branch needs to have the same type. The below function does not compile as one branch tries to return an ``Int`` and the other a ``Text``:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- TYPE_ERROR_BEGIN
+  :end-before: -- TYPE_ERROR_END
+
+If we need functions that can return two (or more) types of things we need to encode that in the return type. For two possibilities, it's common to use the ``Either`` type:
+
+.. literalinclude:: daml/daml-intro-9/daml/Main.daml
+  :language: daml
+  :start-after: -- INT_OR_TEXT_BEGIN
+  :end-before: -- INT_OR_TEXT_END
+
+Branching in Actions
+~~~~~~~~~~~~~~~~~~~~
+
+The most common case where this becomes important is inside ``do`` blocks. Say we want to create contract if a condition is met, or we want to create a contract of one type in one case, and of another type in another case. Let's say we have two template types and want to write a function that creates an ``S`` if a condition is met, and a ``T`` otherwise.
```

Comment:
```suggestion
The most common case where this becomes important is inside ``do`` blocks. Say we want to create a contract of one type in one case, and of another type in another case. Let's say we have two template types and want to write a function that creates an ``S`` if a condition is met, and a ``T`` otherwise.
```